### PR TITLE
[python] Fix continued train by reusing the same dataset

### LINKED ISSUE #2899: [bug] need to reset init_score when use the same init_predictor handle, but with different model contents
I want to train a lightgbm model continuously for multiple times. For example, I have a inital lightgbm m(0), and I want to train m(t+1) based on m(t) by adding N more iterations each time. If I've got it right, this could be implemented by use m(t) as the init_model for training m(t+1). This works fine for m(0) -> m(1). However, when I train m(2) based on m(1), I found that the validation score is exactly the same as before, which is illustrated below. 

m(0)
Training until validation scores don't improve for 50 rounds.
[100]	valid_0's multi_logloss: 0.965397
[200]	valid_0's multi_logloss: 0.795085
[300]	valid_0's multi_logloss: 0.705775
[400]	valid_0's multi_logloss: 0.650636
[500]	valid_0's multi_logloss: 0.61173
Did not meet early stopping. Best iteration is:
[500]	valid_0's multi_logloss: 0.61173
m(1)
Training until validation scores don't improve for 50 rounds.
[600]	valid_0's multi_logloss: 0.580155
[700]	valid_0's multi_logloss: 0.552838
[800]	valid_0's multi_logloss: 0.529345
[900]	valid_0's multi_logloss: 0.508125
[1000]	valid_0's multi_logloss: 0.489261
Did not meet early stopping. Best iteration is:
[1000]	valid_0's multi_logloss: 0.489261
m(2)
Training until validation scores don't improve for 50 rounds.
[1100]	valid_0's multi_logloss: 0.580155
[1200]	valid_0's multi_logloss: 0.552838
[1300]	valid_0's multi_logloss: 0.529345
[1400]	valid_0's multi_logloss: 0.508125
[1500]	valid_0's multi_logloss: 0.489261

You can see that the validation score on iteration 1100-1500 is exactly the same as iteration 600-1000, which is quite weird. It seems that m(2) is still trained based on m(0) instead of m(1) according to the output log. So I want to ask that what is the reason of this problem? And what should be the right way to retrain for multiple times? 

Generally I use the following code for retraining multiple times. 
`model = lgb.train({**params}, train_data, valid_sets=valid_data, early_stopping_rounds=50, verbose_eval=100, init_model=model, keep_training_booster=True)`

### PR DESCRIPTION
**Pull request by guolinke**
fix #2899

### CODE REVIEW

File: tests/python_package_test/test_engine.py [hunk 1]
```diff
@@ -613,6 +613,19 @@ def test_continue_train(self):
...
 
+    def test_continue_train_reused_dataset(self):
+        X, y = load_boston(True)
+        params = {
+            'objective': 'regression',
+            'verbose': -1
+        }
+        lgb_train = lgb.Dataset(X, y, free_raw_data=False)
+        init_gbm = lgb.train(params, lgb_train, num_boost_round=20)
+        init_gbm_2 = lgb.train(params, lgb_train, num_boost_round=20,  init_model=init_gbm)
```
Comment (guolinke) [inline tests/python_package_test/test_engine.py]:
```suggestion
        init_gbm_2 = lgb.train(params, lgb_train, num_boost_round=20, init_model=init_gbm)
```

File: tests/python_package_test/test_engine.py [hunk 2]
```diff
@@ -613,6 +613,19 @@ def test_continue_train(self):
...
+        init_gbm_3 = lgb.train(params, lgb_train, num_boost_round=20,  init_model=init_gbm_2)
```
Comment (guolinke) [inline tests/python_package_test/test_engine.py]:
```suggestion
        init_gbm_3 = lgb.train(params, lgb_train, num_boost_round=20, init_model=init_gbm_2)
```

File: tests/python_package_test/test_engine.py [hunk 3]
```diff
@@ -613,6 +613,19 @@ def test_continue_train(self):
...
+        gbm = lgb.train(params, lgb_train, num_boost_round=20,  init_model=init_gbm_3)
```
Comment (guolinke) [inline tests/python_package_test/test_engine.py]:
```suggestion
        gbm = lgb.train(params, lgb_train, num_boost_round=20, init_model=init_gbm_3)
```

File: python-package/lightgbm/basic.py [hunk 4]
```diff
@@ -725,6 +721,20 @@ def __pred_for_csc(self, csc, num_iteration, predict_type):
         if n_preds != out_num_preds.value:
             raise ValueError("Wrong length for predict results")
         return preds, nrow
+
```
Comment (guolinke) [inline python-package/lightgbm/basic.py]:
```suggestion

```

File: python-package/lightgbm/basic.py [hunk 5]
```diff
@@ -842,27 +852,32 @@ def _set_init_score_by_predictor(self, predictor, data, used_indices=None):
...
         num_data = self.num_data()
-        if used_indices is not None:
-            assert not self.need_slice
-            if isinstance(data, string_type):
-                sub_init_score = np.zeros(num_data * predictor.num_class, dtype=np.float32)
-                assert num_data == len(used_indices)
-                for i in range_(len(used_indices)):
+        if predictor is not None:
+            init_score = predictor.predict(data,
+                                        raw_score=True,
```
Comment (guolinke) [inline python-package/lightgbm/basic.py]:
```suggestion
                                            raw_score=True,
```

File: python-package/lightgbm/basic.py [hunk 6]
```diff
@@ -842,27 +852,32 @@ def _set_init_score_by_predictor(self, predictor, data, used_indices=None):
...
-            for i in range_(num_data):
-                for j in range_(predictor.num_class):
-                    new_init_score[j * num_data + i] = init_score[i * predictor.num_class + j]
-            init_score = new_init_score
+                        new_init_score[j * num_data + i] = init_score[i * predictor.num_class + j]
+                init_score = new_init_score
+        elif self.init_score is not None:
+            init_score = np.zeros(self.init_score.shape, dtype=np.float32)
+        else:
+            return
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py]:
```suggestion
            return self
```

File: python-package/lightgbm/basic.py [hunk 7]
```diff
@@ -1375,16 +1390,20 @@ def _set_predictor(self, predictor):
...
+        elif self.data is not None:
+            self._predictor = predictor
+            self._set_init_score_by_predictor(self._predictor, self.data)
+        elif self.used_indices is not None and self.reference is not None and self.reference.data is not None:
+            self._predictor = predictor
+            self._set_init_score_by_predictor(self._predictor, self.reference.data, self.used_indices)
         else:
             raise LightGBMError("Cannot set predictor after freed raw data, "
                                 "set free_raw_data=False when construct Dataset to avoid this.")
+        return self
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py]: Why don't we need to call `self._free_handle()` anymore?

File: tests/python_package_test/test_engine.py [hunk 8]
```diff
@@ -613,6 +613,19 @@ def test_continue_train(self):
...
+        params = {
+            'objective': 'regression',
+            'verbose': -1
+        }
+        lgb_train = lgb.Dataset(X, y, free_raw_data=False)
+        init_gbm = lgb.train(params, lgb_train, num_boost_round=20)
+        init_gbm_2 = lgb.train(params, lgb_train, num_boost_round=20, init_model=init_gbm)
+        init_gbm_3 = lgb.train(params, lgb_train, num_boost_round=20, init_model=init_gbm_2)
+        gbm = lgb.train(params, lgb_train, num_boost_round=20, init_model=init_gbm_3)
+        self.assertEqual(gbm.current_iteration(), 80)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py]: Let's save CI time.
```suggestion
        init_gbm = lgb.train(params, lgb_train, num_boost_round=5)
        init_gbm_2 = lgb.train(params, lgb_train, num_boost_round=5, init_model=init_gbm)
        init_gbm_3 = lgb.train(params, lgb_train, num_boost_round=5, init_model=init_gbm_2)
        gbm = lgb.train(params, lgb_train, num_boost_round=5, init_model=init_gbm_3)
        self.assertEqual(gbm.current_iteration(), 20)
```

File: python-package/lightgbm/basic.py (hunk 7 above)
Comment (guolinke) [reply python-package/lightgbm/basic.py]: As the init_score could be updated independently, without changing features. I think we don't need free_handle.

**Pull request merged**